raft/node.go

```diff
@@ -114,11 +114,25 @@ type Node interface {
 	ProposeConfChange(ctx context.Context, cc pb.ConfChange) error
 	// Step advances the state machine using the given message. ctx.Err() will be returned, if any.
 	Step(ctx context.Context, msg pb.Message) error
+
 	// Ready returns a channel that returns the current point-in-time state.
-	// Users of the Node must call Advance after applying the state returned by Ready.
+	// Users of the Node must call Advance after retrieving the state returned by Ready.
+	//
+	// NOTE: No committed entries from the next Ready may be applied until all committed entries
+	// and snapshots from the previous one have finished.
 	Ready() <-chan Ready
-	// Advance notifies the Node that the application has applied and saved progress up to the last Ready.
+	// Advance notifies the Node that the application has saved progress up to the last Ready.
 	// It prepares the node to return the next available Ready.
+	//
+	// In normal case, application should call Advance after it applies the entries in last Ready.
+	// When the last Ready contains a snapshot, the application might take a long time to apply
+	// the snapshot data. To continue receiving Ready without blocking raft progress, it is safe(*)
+	// to call Advance before finish applying the last ready. But the applied index in your application
+	// will be behind the applied index that tracks by Node.
+	//
+	// (*) NOTE: If a follower changes its state to candidate, it MUST wait for finishing all previous
```

Comment:
This is subtle, and I think we were getting it wrong before (I didn't realize this was a requirement). What exactly does "changes its state to candidate" mean? A node may change its state to candidate on a `Tick()` and then back to follower when it receives a new message; the application may not observe this if it doesn't call `Ready` during that time. Does that still count, or is the problem only when `Ready.SoftState.RaftState == StateCandidate`? 

I think it would be better for the comment to start with the safe option: "the application should generally apply all entries before calling `Advance`. However, as an optimization, the application may call `Advance` while it is applying the commands if the following conditions hold:...  This is most useful when applying the snapshot, which might take a long time and during which the application should continue to respond to heartbeats."